src/smc-webapp/project/ask-filename.tsx

```diff
@@ -0,0 +1,106 @@
+import { Component, React, Rendered } from "../app-framework";
+const { default_filename } = require("../account.coffee");
+const { file_options } = require("../editor");
+const {
+  Col,
+  Row,
+  ButtonToolbar,
+  ControlLabel,
+  Button,
+  Form
+} = require("react-bootstrap");
+const { SearchInput } = require("../r_misc");
+const { IS_TOUCH } = require("../feature");
+
+interface Props {
+  actions: any;
+  ext_selection: string;
+  current_path: string;
+}
+interface State {
+  new_name: string;
+}
+
+export class AskNewFilename extends Component<Props, State> {
+  displayName = "ProjectFiles-AskNewFilename";
+  constructor(props) {
+    super(props);
+    this.state = {
+      new_name: default_filename()
+    };
+  }
+
+  componentDidMount() {
+    this.setState({ new_name: default_filename() });
+  }
+
+  cancel = (): void => {
+    this.props.actions.setState({ ext_selection: undefined });
```

Comment:
Make a new action instead of directly setting state.

Reply:
> Make a new action instead of directly setting state.

Based on my experience writing a lot of code at this point, I definitely don't necessarily agree...  I've overall found it much better to just have a good definition of the data in the store, and found putting lots of actions in as wrappers around this to be not necessary, annoying, dangerous (since they can do other things and change, so you're never really sure what state is changing).   I ended up using setState directly a lot more in the newer frame-tree code I wrote, and was way happier with it.     I'm just speaking from experience, and am surprised.